### DISCUSSION TITLE
USB MIDI Device "Co-processor" IO board for RPi

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#480] [Show and tell]
I've put together a USB MIDI Device IO board for the Raspberry PI based on the Seeed Studio XIAO SAMD21.  It is basically a variation on the Pico application linked to from the wiki.

Still a bit of a WIP, but it generally works.  Design and build notes are here: https://diyelectromusic.wordpress.com/2023/04/16/xiao-pi-usb-midi-device-pcb/

Here is a short video of it sort of in action (the MIDI file isn't ideal - I should have spent a bit more time making it more suitable for MiniDexed really) - but it just needs USB power to the Pi, USB link to the XIAO and audio output.

At present it ignores MIDI real time messages and SysEx, so the code needs a bit more developing really, but its a start.

Also, it won't play nicely with another serial MIDI interface directly connected to the Pi, hence the video shows a version of my SSD1306 MiniDexed io board with no MIDI itself.

Kevin

https://www.youtube.com/watch?v=JqoAcjEZ7S0

**Comment by probonopd** [comment]
Hi Kevin, is there a special reason you picked this comparatively less popular board over e.g., a RPi Pico?

**Reply by diyelectromusic** [reply]
Is it less popular?  I've seen quite a lot of projects use them :) Its a very popular form factor with Seeed and Adafruit!

But no, not really.  I've been meaning to do things with the XIAO for a while, so just ran with it.

I think about a Pico, but I didn't want to be messing with the Picos "lots of dev environments, but none of them do everything" thing - I'd want multiple serial ports, USB device functionality, PIO... so that is really getting down to just using the C/C++ SDK (Arduino isn't very robust, Micropython doesn't do USB HID MIDI, CircuitPython isn't so good for PIO, and I hate it's MIDI implementation!).  It's also quite big but I can move to a RP2040 board with a XIAO footprint relatively easily if I wanted to in the future.

The SAMD21 version has three hardware serial ports available to the arduino environment - I needed two here - and USB support is pretty secure.  And if you'd rather you can have an ESP32 (S2, S3 or even the RISK-V C3) version and might even get away with an Adafruit QT Py variant...

So, basically, small, easy to code for what I wanted, several hardware serial ports and USB, relatively cheap, very common MCU architecture (SAMD21)...

But mostly "just because" :)

Kevin

**Reply by probonopd** [reply]
Well, of course I have no scientific proof but personally I have a couple of RPi Picos and ESP32XXs lying around, so maybe I am biased ;-)

**Reply by diyelectromusic** [reply]
I stopped counting my microcontrollers many years ago.... good job really!  Let's just say when the semiconductor shortage hit, I didn't really notice too much :)

**Reply by scuzzphut** [reply]
Kevin , if you have any Teensy 3.2s in your stock you could make a fortune on the Ornament & Crime black market.

**Reply by diyelectromusic** [reply]
Nah, just a 3.6 and a 4.0...
